cpu/atmega_common/periph/rtt.c

```diff
@@ -0,0 +1,304 @@
+ * incremented by the overflow ISR. This MUST BE marked volatile for two
+ * reasons: to guarantee ordering with other register operations, and to
+ * force redundant access to occur (in case it was changed by the ISR). */
+static volatile uint16_t ext_cnt;
+
+static inline void _asynch_wait(void)
+{
+    /* Wait until all busy flags clear. According to the datasheet,
+     * this can take up to 2 positive edges of TOSC1 (32kHz). */
+    while (ASSR & ((1 << TCN2UB) | (1 << OCR2AUB) | (1 << OCR2BUB)
+                | (1 << TCR2AUB) | (1 << TCR2BUB))) {}
+}
+
+static inline uint8_t _rtt_div(uint16_t freq)
+{
+    switch (freq) {
+    case 32768: return 0x1;
+    case  4096: return 0x2;
+    case  1024: return 0x3;
+    case   512: return 0x4;
+    case   256: return 0x5;
+    case   128: return 0x6;
+    case    32: return 0x7;
+    default   : assert(0);
+    }
+}
+
+void rtt_init(void)
+{
+    DEBUG("Initializing RTT\n");
+
+    rtt_poweron();
+
+    /*
+     * From the datasheet section "Asynchronous Operation of Timer/Counter2"
+     * p148 for ATmega1284P.
+     * 1. Disable the Timer/Counter2 interrupts by clearing OCIE2x and TOIE2.
+     * 2. Select clock source by setting AS2 as appropriate.
+     * 3. Write new values to TCNT2, OCR2x, and TCCR2x.
+     * 4. To switch to asynchronous: Wait for TCN2UB, OCR2xUB, TCR2xUB.
+     * 5. Clear the Timer/Counter2 Interrupt Flags.
+     * 6. Enable interrupts, if needed
+     */
+
+    /* Disable all timer 2 interrupts */
+    TIMSK2 = 0;
+
+    /* Select asynchronous clock source */
+    ASSR = (1 << AS2);
+
+    /* Set counter to 0 */
+    TCNT2 = 0;
+
+    /* Reset compare values */
+    OCR2A = 0;
+    OCR2B = 0;
+
+    /* Reset timer control */
+    TCCR2A = 0;
+
+    /* 32768Hz / n */
+    TCCR2B = _rtt_div(RTT_FREQUENCY);
+
+    /* Wait until not busy anymore */
+    DEBUG("RTT waits until ASSR not busy\n");
+    _asynch_wait();
+
+    /* Clear interrupt flags */
+    /* Oddly, this is done by writing ones; see datasheet */
+    TIFR2 = (1 << OCF2B) | (1 << OCF2A) | (1 << TOV2);
+
+    /* Enable 8-bit overflow interrupt */
+    TIMSK2 |= (1 << TOIE2);
+
+    DEBUG("RTT initialized\n");
+}
+
+void rtt_set_overflow_cb(rtt_cb_t cb, void *arg)
+{
+    /* Make non-atomic write to callback atomic */
+    unsigned state = irq_disable();
+
+    rtt_state.overflow_cb = cb;
+    rtt_state.overflow_arg = arg;
+
+    irq_restore(state);
+}
+
+void rtt_clear_overflow_cb(void)
+{
+    /* Make non-atomic write to callback atomic */
+    unsigned state = irq_disable();
+
+    rtt_state.overflow_cb = NULL;
+    rtt_state.overflow_arg = NULL;
+
+    irq_restore(state);
+}
+
+uint32_t rtt_get_counter(void)
+{
+    unsigned state;
+    uint32_t now;
+
+    /* Make sure it is safe to read TCNT2, in case we just woke up */
+    DEBUG("RTT sleeps until safe to read TCNT2\n");
+    TCCR2A = 0;
+    _asynch_wait();
+
+    state = irq_disable();
+    now = (ext_cnt << 8) | TCNT2;
```

Comment:
The problem here is that TCNT2 could have just overflown, but ext_cnt is not yet updated. You need to check for the IRQ flag after reading TCNT2 but before restoring IRQs. If the IRQ flag was set, you need to add 1 to the upper 16 bits. If TCNT2 was 255, the lower 8 bits of `now` need additionally be set to `0`.

Have you checked what happens if `volatile` is dropped from `ext_cnt`. Without nested interrupts it should be fine to use it as normal memory from the ISR. And here you are using `irq_disable()` and `irq_restore()`, which serves as a memory barrier. There *should* now no longer be a need for `volatile`.

Reply:
I thought `volatile` was also needed if a variable could change 'outside the normal program flow', e.g. when a variable is only written in an ISR and the compiler sees it never being called, it might assume it's never written at all.

But then again that would be true for shared libraries and anything using threads.
The first examples for the `volatile` keyword explicitly list variables from interrupts, but that might also be a large cargo cult.

Reply:
It's certainly not used by the network drivers :laughing:

Reply:
It's indeed cargo cult. But embedded C developers sadly just never got the message.

Some nice links on this:

- [Jonathan Corbet et al. - Linux Kernel Documentation - Why the “volatile” type class should not be used](https://www.kernel.org/doc/html/latest/process/volatile-considered-harmful.html)
- [John Regehr - Nine ways to break your systems code using volatile](https://blog.regehr.org/archives/28)
- [Arch D. Robison - Volatile: Almost Useless for Multi-Threaded Programming](https://web.archive.org/web/20181124154026/https://software.intel.com/en-us/blogs/2007/11/30/volatile-almost-useless-for-multi-threaded-programming)

(In case of missing barrier, `volatile` seems to fix the bug. But it doesn't, as it doesn't provide the memory ordering guarantees and the atomic access guarantee. It just replaces a bug triggered always by a bug triggered only very rarely...)